### ISSUE TITLE
prov/sockets: Under heavy load provider can mis-direct messages

### ISSUE THREAD

**Issue by jsquyres** [ofiwg/libfabric#2291]
# Overview

When the sockets provider is under heavy load, messages destined for peer A sometimes end up at peer B.

This problem is seen at the libfabric HEAD as of today (25 Aug 2016): 986d5f47f9435938bb459e8b2569be9288a5a1ce (a.k.a. v1.3.0-662-g986d5f4).

# Reproducer

I have a reproducer that sends and receives messages between multiple "servers" and "clients" using libfabric API calls, specifically forcing the use of the sockets provider:

https://github.com/jsquyres/random-libfabric-tests/tree/master/abuse

## Reproducer behavior

This reproducer is mimicking the behavior of a larger, more complex internal application that exhibits a similar message passing traffic pattern.  Specifically:

* There are N server processes (where N is a "small" number -- e.g., single digits)
* There are M client processes (where `M>=N`; M can be anywhere from `(N+1)` to `(very large)`)
* All communication uses EP_RDM endpoints

### Clients

After receiving all the server addresses via MPI, clients loop forever doing this:
* 5% of the time:
   * There is code to destroy the client's EP and re-initialize all of the client's libfabric data structures.
   * However, I was running into problems with this code, so it is currently commented out -- so 5% of the time, the loop iteration is a no-op.
* 25% of the time:
  * Randomly select a server.
  * If this client is "connected" to this server, "disconnect" from that server.  Meaning:
    * Send a "DISCONNECT" message to the server
    * Remove this server from internal data structures
    * Remove this server from the AV
* The rest of the time:
  * Randomly select a server.
  * If this client is not already "connected" to this server:
     * Send a "CONNECT" message to the server with some meta information about this client
     * Wait for "CONNECT_ACK" message back from the server
  * Send a "SOLICIT_RDMA" message to the server
  * Server will RDMA a message back to me
  * Server will then send a "RDMA_SENT" message back to me

### Servers

After sending their addressing data to all clients (via MPI), servers loop forever waiting for completions:

* If it's a send completion, just make sure the send completed ok
* If it's an RDMA write completion, just make sure the RDMA write completed ok
* If it's a receive completion, it will be one of three incoming messages:
   * CONNECT: from a client, sending a bunch of meta data about itself.  Send back a CONNECT_ACK message.
   * DISCONNECT: from a client, telling us to forget everything about that client (and remove it from the AV)
   * SOLICIT_RDMA: server sends an RDMA message back to the client, and then send a RDMA_SENT message back to the client indicating that the RDMA completed.  **NOTE:** The ordering of messages should guarantee that the RDMA_SENT message does not arrive until the RDMA write has actually arrived at the client.

## Failure behavior

In all cases, both the client and the server add some meta data in the beginning of all messages sent indicating to whom the message is intended.

The failure mode that I see is that server A sometimes receives a message that contains meta data indicating that the message was actually sent to server B.

## Reproducer code

Some helpful tips about the reproducer code:

1. It's C++, but really, it's mostly C code -- it's only C++ so that I could use the STL.
1. The use of MPI is really only for bootstrapping:
   * You run X processes per server, where `X>1`.
   * The first MPI process on each server is designated the "server" -- all other MPI processes on that server are designated "clients".
   * Once the servers open up EP_RDM endpoints, they send their fi_getname() addresses to all the clients via MPI API calls.
1. Additionally, the MPI_Abort() function is used to kill the whole job (i.e., all clients and servers) when an inconsistency is found in the message data (i.e., anyone receives a message that was supposed to be sent to anyone else).

## Local customizations

The reproducer has two things that you might/will need to edit:

1. If you're not using Open MPI, you'll need to edit how the clients figure out that they are the first MPI process on a given node (other MPI's have implementation-specific ways to do this, but I don't know what they are).  See lines 130-139 for the details.
2. The `Makefile` is hard-coded to my own paths; it's not a generalized `configure`-based setup or anything like that.

# Running the test

I have been able to exhibit the bad behavior:

* On networks as slow as 1Gbs (i.e., it doesn't require 10, 40, or 100Gbps networks).
* With 6 nodes and 18 MPI processes (i.e., 1 server and 2 clients per node).
   * YMMV
* With Open MPI:
  * I ensure that Open MPI is not using libfabric, just to be fully sure that it isn't interfering with the test.
  * I further limited MPI's message passing to a single TCP network, just to limit the variables under test.
  * For example:

```
$ mpirun --host a:3,b:3,c:3 -np 18 --map-by node \
    --mca btl tcp,vader,self \
    --mca btl_tcp_if_include 10.0.0.0/16 \
    ./abuse
```

**Comment by shantonu** [comment]
Thanks Jeff for the meticulous description. I'll look into the issue.

**Comment by jsquyres** [comment]
@bturrubiates Just had an excellent suggestion to me: remove the call to `fi_av_remove()` in the client.  Guess what?  The problem goes away.

This will likely significantly help narrow down the problem...

**Comment by jsquyres** [comment]
I notice that there's no locking and/or reference counting in `sock_av.c`.  Can there be interactions with the progress thread(s) and an av table, potentially resulting in stale information be used by the progress thread?

**Comment by shantonu** [comment]
I was also thinking if the garbage collection of the av addresses was causing the problem. There could a synchronization issue here. I guess, if you remove f769012df84b7ad0fe5905a79973a0d0733566e0 commit which adds the garbage collection (without removing fi_av_remove), you won't see the problem.

Now I understand the problem better. Will investigate.

**Comment by jsquyres** [comment]
You're right -- reverting https://github.com/ofiwg/libfabric/commit/f769012df84b7ad0fe5905a79973a0d0733566e0 seems to make it work.  This is a little surprising, because I have seen this problem for a little while now -- definitely before https://github.com/ofiwg/libfabric/commit/f769012df84b7ad0fe5905a79973a0d0733566e0.

That being said, when I run this against libfabric v1.3.0, it runs for a little while, but then it eventually assert() fails because `fi_av_insert()` fails (perhaps/probably due to #2226?  I didn't check deeply).

**Comment by jsquyres** [comment]
FWIW, I just noticed that my `ofi_av_test` has odd behavior with the current libfabric HEAD.  The test basically does this:

```c
// create av table of size 32
for (i = 0; i < 128; ++i) {
    fi_av_insert(...);
    fi_av_remove(...);
}
```

Here's the output from the test:

```
Added address 0: 0x0
Added address 1: 0x1
Added address 2: 0x2
Added address 3: 0x3
Added address 4: 0x4
Added address 5: 0x5
...
Added address 29: 0x1d
Added address 30: 0x1e
Added address 31: 0x1f
Added address 32: 0x0
Added address 33: 0x0
Added address 34: 0x0
...
```

I.e., it counts upward in the returned handles, and starting with the 33rd address inserted, all subsequent returned handles are 0.

Was that intended?

**Comment by shantonu** [comment]
This happens when garbage collection is added. In sockets the addresses are just index to the av_table. It keeps inserting to the table until it hits the default size. This happens till insert number 32. Then for the next insert it tries to find the first invalid index which is index 0 in this case, and inserts to that position. That's why after 32th insert add the addresses are inserted at index 0 and has address 0x0.

**Comment by jsquyres** [comment]
That's a little odd -- it seems to be doing 2 behaviors:

1. *Not* garbage collecting (i.e., returning 0, then 1, then 2, ...etc).
1. Garbage collecting (i.e., returning 0, then 0, then 0, ...etc.).

I don't know if that's worth "fixing" (especially if it's a) correct behavior in the table itself, and b) intended behavior 😄 ), but it did catch me by surprise.  Just wanted to throw my $0.02 in...

**Comment by shantonu** [comment]
I agree it looks a bit odd, but the first scenario was added just to avoid lookup when count < size.  I'll revisit this issue after this bug.

**Comment by jsquyres** [comment]
Awesome sauce; thanks (and no rush -- it's just something I noticed).

**Comment by shantonu** [comment]
@jsquyres, I found an issue with connection lookup after adding garbage collection of av index. We were using an av index map for looking up connection, and it didn't know that the address is removed and the index is being reused. It was returning the previously established connection (which was for peer B). Added a fix that cleans up previous connection and returns a new one for the newly added address (peer A) in the av table. I verified it with the reproducer and fabtests. Could you please verify from your end? Here is the branch https://github.com/shantonu/libfabric/tree/pr/sockets-av.

Thanks again for creating the awesome reproducer to detect the erroneous behavior. It was really helpful to find the root cause!

**Comment by jsquyres** [comment]
Actually, I get a new failure mode when I run with that branch.

I'm running on 2 nodes, with 8 processes (i.e., 1 "server" process and 3 "client" processes on each node).

After running correctly for a while, no more completions occur -- everyone starts blocking in timeouts:

```
ivy02:32444:MCW 7:CLIENT: wokeup from epoll due to timeout
ivy02:32442:MCW 3:CLIENT: wokeup from epoll due to timeout
ivy02:32443:MCW 5:CLIENT: wokeup from epoll due to timeout
ivy01:27522:MCW 4:CLIENT: wokeup from epoll due to timeout
ivy01:27521:MCW 2:CLIENT: wokeup from epoll due to timeout
ivy02:32441:MCW 1:SERVER: wokeup from epoll due to timeout
ivy01:27520:MCW 0:SERVER: wokeup from epoll due to timeout
ivy01:27523:MCW 6:CLIENT: wokeup from epoll due to timeout
ivy02:32444:MCW 7:CLIENT: wokeup from epoll due to timeout
ivy02:32442:MCW 3:CLIENT: wokeup from epoll due to timeout
ivy02:32443:MCW 5:CLIENT: wokeup from epoll due to timeout
ivy01:27522:MCW 4:CLIENT: wokeup from epoll due to timeout
```

From gdb attaching, it looks like the server is waking up due to timeouts from `epoll_wait()` -- it's not waking up from any completions (i.e., no new completions are arriving).  And therefore the clients are blocking while waiting for the CONNECT_ACK message from the server.  Here's the last bunch of things that one of the servers did before it entered an endless wakeup-from-epoll-with-no-completions loop:

```
ivy01:29221:MCW 0:SERVER: Sending to fi_addr 0x1
ivy01:29221:MCW 0:SERVER: Sent CONNECT_ACK back to 10.0.20.239:50301, fi_addr 0x1
ivy01:29221:MCW 0:SERVER: Completed a send
ivy01:29221:MCW 0:SERVER: Got SOLICIT_RDMA from 10.0.20.239:50301 (MCW rank 6), seq=899
ivy01:29221:MCW 0:SERVER: RDMA writing to 10.0.20.239:50301 (fi_addr 0x1, MCW rank 6), dest buffer=0x2aaabf09ed90 - 0x2aaabf273326 (len=1918358), seq=3876, dest key=0
ivy01:29221:MCW 0:SERVER: Fill header: from mcw=0, from ip=10.0.20.239:50572
ivy01:29221:MCW 0:SERVER: Sending to fi_addr 0x1
ivy01:29221:MCW 0:SERVER: Sent RDMA_SENT back to 10.0.20.239:50301, fi_addr 0x1
ivy01:29224:MCW 6:CLIENT: Got MSG_RDMA_SENT: server wrote 1918358 bytes to 0x2aaabf09ed90
ivy01:29221:MCW 0:SERVER: Completed a send
ivy01:29221:MCW 0:SERVER: Got SOLICIT_RDMA from 10.0.20.239:50301 (MCW rank 6), seq=900
ivy01:29221:MCW 0:SERVER: RDMA writing to 10.0.20.239:50301 (fi_addr 0x1, MCW rank 6), dest buffer=0x2aaabf075730 - 0x2aaabf22ef93 (len=1808483), seq=3878, dest key=0
ivy01:29221:MCW 0:SERVER: Fill header: from mcw=0, from ip=10.0.20.239:50572
ivy01:29221:MCW 0:SERVER: Sending to fi_addr 0x1
ivy01:29221:MCW 0:SERVER: Sent RDMA_SENT back to 10.0.20.239:50301, fi_addr 0x1
ivy01:29224:MCW 6:CLIENT: Got MSG_RDMA_SENT: server wrote 1808483 bytes to 0x2aaabf075730
ivy01:29221:MCW 0:SERVER: Completed a send
ivy01:29221:MCW 0:SERVER: Got SOLICIT_RDMA from 10.0.20.239:50301 (MCW rank 6), seq=901
ivy01:29221:MCW 0:SERVER: RDMA writing to 10.0.20.239:50301 (fi_addr 0x1, MCW rank 6), dest buffer=0x2aaabeb5b060 - 0x2aaabef25d6a (len=3976458), seq=3880, dest key=0
ivy01:29221:MCW 0:SERVER: Fill header: from mcw=0, from ip=10.0.20.239:50572
ivy01:29221:MCW 0:SERVER: Sending to fi_addr 0x1
ivy01:29221:MCW 0:SERVER: Sent RDMA_SENT back to 10.0.20.239:50301, fi_addr 0x1
ivy01:29224:MCW 6:CLIENT: Got MSG_RDMA_SENT: server wrote 3976458 bytes to 0x2aaabeb5b060
ivy01:29221:MCW 0:SERVER: Completed a send
ivy01:29221:MCW 0:SERVER: Got MSG_DISCONNECT from 10.0.20.239:50301 (MCW rank 6), seq=902
ivy01:29221:MCW 0:SERVER: fi_av_remove of 10.0.20.239:50301 --> fi_addr_t 0x1
ivy01:29221:MCW 0:SERVER: wokeup from epoll due to timeout
ivy01:29221:MCW 0:SERVER: wokeup from epoll due to timeout
ivy01:29221:MCW 0:SERVER: wokeup from epoll due to timeout
```

**Comment by shantonu** [comment]
Hmm, I guess, I didn't run long enough. Will keep digging.

**Comment by jsquyres** [comment]
FWIW, it consistently fails for me at about 20 seconds into the run.  Perhaps running on 2 nodes / 8 procs total is some kind of magic combination...?

**Comment by shantonu** [comment]
@jsquyres, I was missing a check in connection lookup. Pushed the fix in https://github.com/shantonu/libfabric/tree/pr/sockets-av. But I get another error ```...but we're already connected``` from ```server.cc``` when ```MSG_CONNECT``` is received. This can happen in the following scenario since there's no ```MSG_DISCONNECT_ACK``` -
```
- client1 sends MSG_DISCONNECT to server1
- client1 removes server1 from it's internal map
- client1 sends MSG_CONNECT to server1
- server1 receives MSG_CONNECT before receiving MSG_DISCONNECT
```
I added ```MSG_DISCONNECT_ACK``` in the reproducer and it works fine with my branch with 2 nodes and 8 processes. Could you please try this this patch? [disconnect-ack.txt](https://github.com/ofiwg/libfabric/files/470607/disconnect-ack.txt)

**Comment by jsquyres** [comment]
Forgive me if this is a dumb question, but how can the MSG_CONNECT arrive at the server before the MSG_DISCONNECT?  Doesn't the EP_RDM endpoint enforce send/receive ordering?

**Comment by jsquyres** [comment]
FWIW, I also tried this patch (*instead* of your disconnect-ack.txt patch):

```diff
diff --git a/abuse/main.cc b/abuse/main.cc
index 5fd1557..417b0a1 100644
--- a/abuse/main.cc
+++ b/abuse/main.cc
@@ -407,6 +407,9 @@ void setup_ofi_device(void)
     // Make a fabric from first info returned
     // Ensure we're in "basic" MR mode!
     fidev.info->domain_attr->mr_mode = FI_MR_BASIC;
+    // Make sure we specify we want message ordering
+    fidev.info->tx_attr->comp_order |= FI_ORDER_SAS;
+    fidev.info->rx_attr->comp_order |= FI_ORDER_SAS;
     ret = fi_fabric(fidev.info->fabric_attr, &fidev.fabric, NULL);
     if (0 != ret) {
         error("fi_fabric failed");
```

but I see the same behavior you mentioned: the client sends DISCONNECT, then CONNECT, and the server somehow gets the CONNECT before the DISCONNECT (or perhaps doesn't receive the DISCONNECT -- can't tell easily because my test aborts when it gets the CONNECT unexpectedly).

**Comment by shefty** [comment]
I'm trying to follow this thread.  Does the original problem still exist?  I.e. the AV is still broken?

**Comment by jsquyres** [comment]
I think @shantonu has fixed the original + followup problem.

My latest question is whether EP_RDM is guaranteed to preserve message ordering.  If it is -- or even if I need to specify `FI_ORDER_SAS` (per the patch above) -- then we may have discovered a new bug in the sockets provider...?

**Comment by shefty** [comment]
EP_RDM is not supposed to guarantee message ordering unless explicitly requested by the app.  This way individual packets can take separate paths through the fabric, allow for out of order ACK processing, etc.

**Comment by shantonu** [comment]
Sockets provider should receive the messages in order since it's using tcp communication. Sean and I went over the latest issue and got some ideas. I'll update you when I have a fix.

**Comment by jsquyres** [comment]
@shantonu Sweet; thanks!

**Comment by jsquyres** [comment]
@shefty Are you referring to `FI_ORDER_SAS`?  That seems to be the only `FI_ORDER_*` flag that refers to send semantics (that are unrelated to RMA semantics).

**Comment by shefty** [comment]
I was simply stating that RDM endpoints do not guarantee any ordering unless specified (e.g. FI_ORDER_SAS).  The sockets provider may always guarantee ordering based on how it's implemented however.

We've identified an issue with the sockets provider that I think will result in messages being sent to the wrong peer.  Shantonu is working on developing a fix.

**Comment by jsquyres** [comment]
@shefty Ah, I misunderstood the above comment, then.  Are you saying that your inspection found (another) place where a message sent to A could mistakenly be sent to B instead of C?  I ask because I interpreted the comments above to mean that that you were fixing an out-of-order issue (which, admittedly, is surprising because of the use of TCP for the transport, but I suppose it could happen if the sockets provider switches between writing down the TCP sockets and queueing, etc.).

Not that this really matters -- I'm just curious.  😄

**Comment by shefty** [comment]
> Are you saying that your inspection found (another)
> place where a message sent to A could mistakenly be sent to B instead
> of C?

Yes

I'm not sure if out-or-order is actually occurring, versus the peers receiving data that wasn't intended for them, which we're interpreting as out-of-order messages.

**Comment by shantonu** [comment]
@jsquyres, I have updated the my branch (https://github.com/shantonu/libfabric/tree/pr/sockets-av) based on the discussion with @shefty. I think there's no mis-directed message now. But still I see the following error from the reproducer-
```
MCW 0:SERVER: Got MSG_CONNECT from 192.168.122.15:34490 (MCW rank 3), seq=62
MCW 0:SERVER:line 119:ERROR: ...but we're already connected
```
But I think this can happen with current workflow of the reproducer. Here's the scenario-

1. client1 sends MSG_DISCONNECT to server1
2. client1 calls fi_av_remove with server1's address
3. client1 sends MSG_CONNECT to server1
4. server1 receives MSG_CONNECT but doesn't get MSG_DISCONNECT

Since ```fi_av_remove``` is called right after sending MSG_DISCONNECT, the behavior of the operations that are in progress which uses the removed address, is not defined. So the above error can happen.

**Comment by jsquyres** [comment]
Yes, you are right -- the client needs to wait for the send completion of MSG_DISCONNECT before calling fi_av_remove.  When I fixed that, the problem goes away.  Yay!

I'm seeing a memory leak on the server side of my abuse program.  I'm not yet sure if it's a bug in my test program or if it's a leak in libfabric / the sockets provider.  Regardless, that is not related to this issue, so I think you should PR your fixes in to libfabric and mark them as "fixes #2291".  I'll open a new issue if I find a memory leak.

**Comment by shantonu** [comment]
Great! Thanks a lot for verifying! I'll submit the PR then :)
